HACS support + latest noaa_coop version

rsnodgrass opened the pull request:
* add support for HACS (plus updating README.md with installation instructions)
* editing of README and examples
* bug fix for self.attr missing self reference
* error message bug fix
* bump to latest version of noaa_coop
* ran black, refurb, isort Python code formating/standard tools

rsnodgrass committed:
update

rsnodgrass committed:
updated for HACS

rsnodgrass committed:
updated example

rsnodgrass committed:
fixed error message formating when there is a tuple returned for status_code

rsnodgrass committed:
updates

rsnodgrass committed:
updated log

rsnodgrass committed:
fixed bug where returning attr instead of self.attr

rsnodgrass committed:
added symlink to appease HACS directory structure

rsnodgrass committed:
updated docs

rsnodgrass committed:
pretty format the yaml

rsnodgrass committed:
Merge branch 'master' into master

rsnodgrass committed:
update

rsnodgrass committed:
merged jshufro changes

rsnodgrass committed:
Merge branch 'master' of github.com:rsnodgrass/home_assistant_noaa_tides into jshufro-master

rsnodgrass committed:
Merge branch 'jshufro-master'

rsnodgrass committed:
updated manifest, bumped noaa_coop version dependency, bumped version

rsnodgrass committed:
update manifests

rsnodgrass committed:
updated docs

rsnodgrass committed:
fix quoting

rsnodgrass committed:
ran black, refurb, isort code cleanup tools

rsnodgrass committed:
added project code cleanup config

rsnodgrass committed:
minor doc format

rsnodgrass commented:
Checking on this merge.

rsnodgrass committed:
update

rsnodgrass commented:
Any chance this would get merged into your official version so I don't have to maintain a fixed version for others?

jshufro commented:
> @jshufro Any chance this would get merged into your official version so I don't have to maintain a fixed version for others?

Main issue I see is the breaking change from renaming the path to support HACS. Many people just clone this repo under custom_components.

DanceMore commented:
but with HACS, people can install it and get updates without having to manage `git` repos at all...

I think I subscribed to this issue because I wanted to use the plugin, but I'm already managing so much software that I actually need package managers like HACS (and docker, and others) to manage the sprawl.

jshufro commented:
I'm not opposed to it, just weighing the pain against the benefits.

I think it'll make peoples' lives easier long term. I'll just need to post on the forums and update the readme and all that to make the breaking change extremely well documented.

Some people don't use HACS, and it will certainly make their lives harder- now they have to `git pull` and `cp -R` the `noaa_tides` folder out.

DanceMore commented:
I'm an SRE by trade and one of my biased opinions is "git is not a deployment tool".

You probably use git in your packaging workflows to make "Artifacts" that endusers and sysadmins install. These are ideally some form of Package Managed Object, be it `.deb`, `.rpm`, Docker images, Pypi, Rubygem, NPM, et al, or even HACS :)

Deploying directly from git is acceptable during development; heck, it's often the only choice when you're churning code and need to test it constantly... but it's hard to turn into a robust long-term solution. As you've already seen, managing paths and subdirectories can get annoying for the needs of Developers vs Deployers. It gets even more bulky and complicated if you start adding test suites, `.github/` automations, or documentation that needs a toolchain to generate. Life gets very confusing when you get bug reports from someone that isn't running an "official released version number" and no git commit-ish that you can find in your history .....

jshufro commented:
Python doesn't exactly need a build pipeline.

I'd prefer to have set the repo up in a hacs-compatible way from the start, but not having done that, I don't have the luxury of blindly reorganizing things without considering what might break. As an SRE I am sure you are aware.

jshufro commented:
@rsnodgrass looking at the options for manifest.json, it seems like setting `filename` should remove the need to change the directory structure- can you test that out?

![image](https://github.com/jshufro/home_assistant_noaa_tides/assets/116244/5299e964-6d4c-4af5-b21e-562107abd4f1)

https://hacs.xyz/docs/publish/start/#hacsjson

If that doesn't work, i'd be more open to setting `content_in_root` and pulling the .py files up into the project root directory, which would let people simply `git clone git@github.com:jshufro/home_assistant_noaa_tides.git noaa_tides` in the custom_components path if they don't want to use HACS

rsnodgrass commented:
@jshufro since there is only a single sensor.py, the filename approach just might work. I don't know how it works with the manifest.json file.

Another solution is to fork this into a new repo jschufro/hass-noaa-tides, add that to HACS, and then add warnings to the existing one that users should switch to the "new" one.

rsnodgrass committed:
moved files to root and also updates hacs.json to use filename reference for testing

jshufro commented on hacs.json:
```diff
@@ -0,0 +1,6 @@
+    "name": "NOAA Tides and Currents",
+    "domains": ["sensor"],
+    "render_readme": true,
+    "filename": "sensor.py"
```
```suggestion
    "filename": "noaa_coops/sensor.py"
```

This would keep the path structure of the repo intact so it's the least likely to break anyone's setup.

If it isn't tenable for whatever reason, we can go with what you have here, and I'll just warn anyone who cloned into custom_components/ that they need to reclone into a subpath called noaa_coops

rsnodgrass replied:
@jshufro No that does not work since there are only two options for HACS path structures:

1. custom_components/<integration_name>

2. all files in the root of the repository AND the 'content_in_root' set to True in hacs.json

There is no other path structures supported.

However, one option might be to use a symlink. This may not work on Windows based installations, but should on any Linux or Mac HA instances. Nope, that still did not pass the HACS structure check for integrations. 

HOWEVER, how about the reverse. Move noaa_tides into custom_components structure like most integrations that use HACS. Then symlink noaa_tides in root to the new path. This would allow people to install this manually still using whatever mechanism they previously did (e.g. clone, etc) *AND* also allow HACS to work.

rsnodgrass commented:
I will test this later this week!    Thanks.   
  
  
  
  
  
  
  

  
  
>   
> On Nov 20, 2023 at 10:20 AM,  <Jacob Shufro ***@***.***)>  wrote:
>   
>   
>   
>
>
>
>   
>
> @jshufro  commented on this pull request.
>
>     
>
> In   hacs.json (https://github.com/jshufro/home_assistant_noaa_tides/pull/22#discussion_r1399581422):
>
>   
>   >  @@ -0,0 +1,6 @@ +{ + "name": "NOAA Tides and Currents", + "domains": ["sensor"], + "render_readme": true, + "filename": "sensor.py"  
>  ⬇️ Suggested change  
>  - "filename": "sensor.py" + "filename": "noaa_coops/sensor.py"  
>   
>
> This would keep the path structure of the repo intact so it's the least likely to break anyone's setup.
>
>   
>
> If it isn't tenable for whatever reason, we can go with what you have here, and I'll just warn anyone who cloned into custom_components/ that they need to reclone into a subpath called noaa_coops
>
>   
>
> —
>  Reply to this email directly,   view it on GitHub (https://github.com/jshufro/home_assistant_noaa_tides/pull/22#pullrequestreview-1740428797), or   unsubscribe (https://github.com/notifications/unsubscribe-auth/AAQY4XA544QMHJ3E7XCW7DDYFONO3AVCNFSM5Z7R4BMKU5DIOJSWCZC7NNSXTPCQOVWGYUTFOF2WK43UKJSXM2LFO45TCNZUGA2DEOBXHE3Q).
>  You are receiving this because you were mentioned.Message ID:  ***@***.***>
>
>   
>

rjbrown99 commented:
Should I be able to test this in HACS via adding the following?

Repository: rsnodgrass/home_assistant_noaa_tides 
Category: integration

Asking as I tried that, but it comes back with an error.

> <Integration rsnodgrass/home_assistant_noaa_tides> Repository structure for master is not compliant

rsnodgrass committed:
updates per PR #22

rsnodgrass committed:
updated to latest pre-commit dependencies

rsnodgrass committed:
fixed noaa_coops to noaa_tides

rsnodgrass committed:
testing using symlinks for HACS support without altering noaa_tides path structure

rsnodgrass committed:
symlink did not pass HACS structure checks, changing the noaa_tides root path to be a link into custom_components

rsnodgrass commented:
> Should I be able to test this in HACS via adding the following?
> 
> Repository: rsnodgrass/home_assistant_noaa_tides Category: integration
> 
> Asking as I tried that, but it comes back with an error.
> 
> > <Integration rsnodgrass/home_assistant_noaa_tides> Repository structure for master is not compliant

Try now, the repo structure should now be compliant.

rsnodgrass commented:
@jshufro see my comment above about adding a symlink from noaa_tides to custom_components/noaa_tides. This should address your earlier concern about keeping the repo structure intact for those who have some sort of manual clone/install process.

rsnodgrass committed:
cleaned up pre-commit-config

rjbrown99 commented:
The updated repo does indeed work with HACS now, thank you.

therealcmj commented:
@jshufro and @rsnodgrass as someone just starting on adding tide data to my HA setup thanks to both of you for this work.

It would be really nice to get @jshufro's component into HACS if only to make it easier for people starting out. If there's anything I could do to help move this along I would happily do it. But I think the most I can do is say "please". And thank you!

rsnodgrass committed:
Modernize to Home Assistant 2025.2+ standards

- Update manifest.json with homeassistant 2025.2.0 requirement, integration_type, config_flow
- Add modern config_flow.py with Selectors for better UX
- Implement DataUpdateCoordinator pattern for centralized data fetching
- Add const.py with typed constants
- Modernize sensor.py with proper type hints (| unions, lowercase generics)
- Add CoordinatorEntity base class for all sensors
- Add proper DeviceInfo with configuration URLs
- Add strings.json and translations/en.json with helpful descriptions
- Add diagnostics.py for troubleshooting
- Update hacs.json with minimum HA version
- Add comprehensive test suite for config flow, coordinator, sensors, diagnostics
- Fix all ruff lint errors
